# Project Title : Marketing Analysis

# 1) Introduction
<hr style="border:1px solid black"> </hr>

Nowadays, the role of "Marketing Analyst" is crucial in the business. Their main job is to help companies or organizations decide which products and services to sell, as well as to which customers and at what price. It is of their concern to study the current marketing conditions, consumer behavior, and competitors' activities.

In this project, we aim to investigate the marketing trend and consumer demographics to determine the best marketing approach to reach the target customers. In this case, we used a relevant open-source dataset available in Kaggle (https://www.kaggle.com/datasets/rodsaldanha/arketing-campaign?datasetId=178248&sortBy=voteCount) for our project.

Our goals for this project are as follows:

1. To create a robust model to predict whether a customer's response(accept or decline) with the given attributes
2. To determine the best features (variables) which highly affect the customer's response
3. Suggest specific marketing strategies, promotions, or marketing materials to the company to maximize their revenue or optimizing marketing campaigns.


# 2) Data Preparation and Analysis of Dataset
<hr style="border:1px solid black"> </hr>

We first start by reading our csv file. Lets import all the libraries needed.

In [21]:
# Importing basic libraries
import numpy as np
import pandas as pd
import seaborn as sb
import matplotlib.pyplot as plt
import matplotlib.style as style
from matplotlib import colors
from matplotlib.colors import ListedColormap
import warnings
sb.set()
warnings.filterwarnings("ignore")

In [22]:
campaigndata = pd.read_csv("marketing_campaign.csv")
campaigndata

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
1      2174        1954  Graduation         Single  46344.0        1   
2      4141        1965  Graduation       Together  71613.0        0   
3      6182        1984  Graduation       Together  26646.0        1   
4      5324        1981         PhD        Married  58293.0        1   
...     ...         ...         ...            ...      ...      ...   
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
0            0    4/9/2012       58       635  ...                  7   
1            1    8/3/2014       38        11  ...                  5   
2            0   21/8/2013       26       426  ...                  4   
3            0   10/2/2014       26        11  ...                  6   
4            0   19/1/2014       94       173  ...                  5   
...        ...         ...      ...       ...  ...                ...   
2235         1   13/6/2013       46       709  ...                  5   
2236         1   10/6/2014       56       406  ...                  7   
2237         0   25/1/2014       91       908  ...                  6   
2238         1   24/1/2014        8       428  ...                  3   
2239         1  15/10/2012       40        84  ...                  7   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
0                0             0             0             0             0   
1                0             0             0             0             0   
2                0             0             0             0             0   
3                0             0             0             0             0   
4                0             0             0             0             0   
...            ...           ...           ...           ...           ...   
2235             0             0             0             0             0   
2236             0             0             0             1             0   
2237             0             1             0             0             0   
2238             0             0             0             0             0   
2239             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
0            0              3         11         1  
1            0              3         11         0  
2            0              3         11         0  
3            0              3         11         0  
4            0              3         11         0  
...        ...            ...        ...       ...  
2235         0              3         11         0  
2236         0              3         11         0  
2237         0              3         11         0  
2238         0              3         11         0  
2239         0              3         11         1  

[2240 rows x 29 columns]

In [23]:
# Checking the dimensions and shape of dataset
print("Data type : ", type(campaigndata))
print("Data dims : ", campaigndata.shape)
print(campaigndata.info())

Data type :  <class 'pandas.core.frame.DataFrame'>
Data dims :  (2240, 29)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds 

We see that the original dataset contain **2240** rows of data entries and **29** variables.

Below show the definitions for all attributes:

1. AcceptedCmp1 - 1 if customer accepted the offer in the 1st campaign, 0 otherwise 
2. AcceptedCmp2 - 1 if customer accepted the offer in the 2nd campaign, 0 otherwise 
3. AcceptedCmp3 - 1 if customer accepted the offer in the 3rd campaign, 0 otherwise 
4. AcceptedCmp4 - 1 if customer accepted the offer in the 4th campaign, 0 otherwise 
5. AcceptedCmp5 - 1 if customer accepted the offer in the 5th campaign, 0 otherwise 
6. Response (target) - 1 if customer accepted the offer in the last campaign, 0 otherwise 
7. Complain - 1 if customer complained in the last 2 years
8. DtCustomer - date of customer’s enrolment with the company
9. Education - customer’s level of education
10. Marital - customer’s marital status
11. Kidhome - number of small children in customer’s household
12. Teenhome - number of teenagers in customer’s household
13. Income - customer’s yearly household income
14. MntFishProducts - amount spent on fish products in the last 2 years
15. MntMeatProducts - amount spent on meat products in the last 2 years
16. MntFruits - amount spent on fruits products in the last 2 years
17. MntSweetProducts - amount spent on sweet products in the last 2 years
18. MntWines - amount spent on wine products in the last 2 years
19. MntGoldProds - amount spent on gold products in the last 2 years
20. NumDealsPurchases - number of purchases made with discount
21. NumCatalogPurchases - number of purchases made using catalogue
22. NumStorePurchases - number of purchases made directly in stores
23. NumWebPurchases - number of purchases made through company’s web site
24. NumWebVisitsMonth - number of visits to company’s web site in the last month
25. Recency - number of days since the last purchase
26. Year Birth - year of birth of customer
27. Z_CostContact - Cost required to contact a customer
28. Z_Revenue - Revenue earned if a customer accepted the offer
29. ID - unique id of a customer

## 2.1)  Data Cleaning
<hr style="border:1px solid black"> </hr>

After much analysis and consideration, we decide to engineer our data for a more accessible analysis.

First of all, we decided to combine `AcceptedCMP1`, `AcceptedCMP2`, `AcceptedCMP3`, `AcceptedCMP4`and`AcceptedCMP5`, as they are indicators that indicate whether a customer will accept the offer in which campaign (before the final campaign). We decided to construct a new variable named `Previous` to indicate whether the customer had accepted the campaign before the last campaign. The value for `Previous` will be 1 for that customer if they had taken the offer in any of the above campaigns, and 0 otherwise.

Furthermore, we also removed the variables `Z_Cost Contact`(the cost needed to contact the customer) and `Z_Revenue` (the revenue earned if the customer accepted the offer). We removed these variables, as the cost to contact a customer and the income earned if each customer accepted the offer are the same throughout. This would not give any insights into predicting whether the customer will accept the offer.

Lastly, we remove the column `ID` as this is an uninformative variable. The removal of these variables is just from the first impression of the datasets, further exploration of the data is needed to obtain the best datasets for machine learning.

In [24]:
# Adding new variables
campaigndata["AcceptedCmp1"] = campaigndata["AcceptedCmp1"] + campaigndata["AcceptedCmp2"] + campaigndata["AcceptedCmp3"] + campaigndata["AcceptedCmp4"]+ campaigndata["AcceptedCmp5"]
campaigndata

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
1      2174        1954  Graduation         Single  46344.0        1   
2      4141        1965  Graduation       Together  71613.0        0   
3      6182        1984  Graduation       Together  26646.0        1   
4      5324        1981         PhD        Married  58293.0        1   
...     ...         ...         ...            ...      ...      ...   
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
0            0    4/9/2012       58       635  ...                  7   
1            1    8/3/2014       38        11  ...                  5   
2            0   21/8/2013       26       426  ...                  4   
3            0   10/2/2014       26        11  ...                  6   
4            0   19/1/2014       94       173  ...                  5   
...        ...         ...      ...       ...  ...                ...   
2235         1   13/6/2013       46       709  ...                  5   
2236         1   10/6/2014       56       406  ...                  7   
2237         0   25/1/2014       91       908  ...                  6   
2238         1   24/1/2014        8       428  ...                  3   
2239         1  15/10/2012       40        84  ...                  7   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
0                0             0             0             0             0   
1                0             0             0             0             0   
2                0             0             0             0             0   
3                0             0             0             0             0   
4                0             0             0             0             0   
...            ...           ...           ...           ...           ...   
2235             0             0             0             0             0   
2236             0             0             0             1             0   
2237             0             1             0             1             0   
2238             0             0             0             0             0   
2239             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
0            0              3         11         1  
1            0              3         11         0  
2            0              3         11         0  
3            0              3         11         0  
4            0              3         11         0  
...        ...            ...        ...       ...  
2235         0              3         11         0  
2236         0              3         11         0  
2237         0              3         11         0  
2238         0              3         11         0  
2239         0              3         11         1  

[2240 rows x 29 columns]

In [25]:
campaigndata.rename(columns = {'AcceptedCmp1':'Previous'}, inplace = True)
campaigndata.pop('AcceptedCmp2')
campaigndata.pop('AcceptedCmp3')
campaigndata.pop('AcceptedCmp4')
campaigndata.pop('AcceptedCmp5')

0       0
1       0
2       0
3       0
4       0
       ..
2235    0
2236    0
2237    0
2238    0
2239    0
Name: AcceptedCmp5, Length: 2240, dtype: int64

In [26]:
# 1 and 0 for all rows 
campaigndata.loc[campaigndata['Previous'] >= 1, 'Previous'] = 1

In [27]:
# Removing the variables mentioned
campaigndata.pop('Z_CostContact')
campaigndata.pop('Z_Revenue')
campaigndata.pop('ID')

campaigndata.head()

Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0        1957  Graduation         Single  58138.0        0         0   
1        1954  Graduation         Single  46344.0        1         1   
2        1965  Graduation       Together  71613.0        0         0   
3        1984  Graduation       Together  26646.0        1         0   
4        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  MntWines  MntFruits  ...  MntSweetProducts  \
0    4/9/2012       58       635         88  ...                88   
1    8/3/2014       38        11          1  ...                 1   
2   21/8/2013       26       426         49  ...                21   
3   10/2/2014       26        11          4  ...                 3   
4   19/1/2014       94       173         43  ...                27   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0            88                  3                8                   10   
1             6                  2                1                    1   
2            42                  1                8                    2   
3             5                  2                2                    0   
4            15                  5                5                    3   

   NumStorePurchases  NumWebVisitsMonth  Previous  Complain  Response  
0                  4                  7         0         0         1  
1                  2                  5         0         0         0  
2                 10                  4         0         0         0  
3                  4                  6         0         0         0  
4                  6                  5         0         0         0  

[5 rows x 22 columns]

We can also make some changes to our datasets. The variable `Year_Birth` can be replace by indicating it as the **Age** of that customer to give us better insights on the datasets. We also see that it would be easier for us to treat the column `Kidhome` and `Teenhome` as the number of children in the household. 

The variable `Dt_Customer` would be further transform into the number of Years becoming a customer.

In [28]:
# Creating a columns age of customer
campaigndata.rename(columns = {'Year_Birth':'Age'}, inplace = True)
campaigndata['Age'] = 2022 - campaigndata['Age']
campaigndata

Age   Education Marital_Status   Income  Kidhome  Teenhome Dt_Customer  \
0      65  Graduation         Single  58138.0        0         0    4/9/2012   
1      68  Graduation         Single  46344.0        1         1    8/3/2014   
2      57  Graduation       Together  71613.0        0         0   21/8/2013   
3      38  Graduation       Together  26646.0        1         0   10/2/2014   
4      41         PhD        Married  58293.0        1         0   19/1/2014   
...   ...         ...            ...      ...      ...       ...         ...   
2235   55  Graduation        Married  61223.0        0         1   13/6/2013   
2236   76         PhD       Together  64014.0        2         1   10/6/2014   
2237   41  Graduation       Divorced  56981.0        0         0   25/1/2014   
2238   66      Master       Together  69245.0        0         1   24/1/2014   
2239   68         PhD        Married  52869.0        1         1  15/10/2012   

      Recency  MntWines  MntFruits  ...  MntSweetProducts  MntGoldProds  \
0          58       635         88  ...                88            88   
1          38        11          1  ...                 1             6   
2          26       426         49  ...                21            42   
3          26        11          4  ...                 3             5   
4          94       173         43  ...                27            15   
...       ...       ...        ...  ...               ...           ...   
2235       46       709         43  ...               118           247   
2236       56       406          0  ...                 0             8   
2237       91       908         48  ...                12            24   
2238        8       428         30  ...                30            61   
2239       40        84          3  ...                 1            21   

      NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0                     3                8                   10   
1                     2                1                    1   
2                     1                8                    2   
3                     2                2                    0   
4                     5                5                    3   
...                 ...              ...                  ...   
2235                  2                9                    3   
2236                  7                8                    2   
2237                  1                2                    3   
2238                  2                6                    5   
2239                  3                3                    1   

      NumStorePurchases  NumWebVisitsMonth  Previous  Complain  Response  
0                     4                  7         0         0         1  
1                     2                  5         0         0         0  
2                    10                  4         0         0         0  
3                     4                  6         0         0         0  
4                     6                  5         0         0         0  
...                 ...                ...       ...       ...       ...  
2235                  4                  5         0         0         0  
2236                  5                  7         1         0         0  
2237                 13                  6         1         0         0  
2238                 10                  3         0         0         0  
2239                  4                  7         0         0         1  

[2240 rows x 22 columns]

In [29]:
campaigndata['ChildrenHome'] = campaigndata['Kidhome'] + campaigndata['Teenhome']
campaigndata.pop('Kidhome')
campaigndata.pop('Teenhome')
campaigndata

Age   Education Marital_Status   Income Dt_Customer  Recency  MntWines  \
0      65  Graduation         Single  58138.0    4/9/2012       58       635   
1      68  Graduation         Single  46344.0    8/3/2014       38        11   
2      57  Graduation       Together  71613.0   21/8/2013       26       426   
3      38  Graduation       Together  26646.0   10/2/2014       26        11   
4      41         PhD        Married  58293.0   19/1/2014       94       173   
...   ...         ...            ...      ...         ...      ...       ...   
2235   55  Graduation        Married  61223.0   13/6/2013       46       709   
2236   76         PhD       Together  64014.0   10/6/2014       56       406   
2237   41  Graduation       Divorced  56981.0   25/1/2014       91       908   
2238   66      Master       Together  69245.0   24/1/2014        8       428   
2239   68         PhD        Married  52869.0  15/10/2012       40        84   

      MntFruits  MntMeatProducts  MntFishProducts  ...  MntGoldProds  \
0            88              546              172  ...            88   
1             1                6                2  ...             6   
2            49              127              111  ...            42   
3             4               20               10  ...             5   
4            43              118               46  ...            15   
...         ...              ...              ...  ...           ...   
2235         43              182               42  ...           247   
2236          0               30                0  ...             8   
2237         48              217               32  ...            24   
2238         30              214               80  ...            61   
2239          3               61                2  ...            21   

      NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0                     3                8                   10   
1                     2                1                    1   
2                     1                8                    2   
3                     2                2                    0   
4                     5                5                    3   
...                 ...              ...                  ...   
2235                  2                9                    3   
2236                  7                8                    2   
2237                  1                2                    3   
2238                  2                6                    5   
2239                  3                3                    1   

      NumStorePurchases  NumWebVisitsMonth  Previous  Complain  Response  \
0                     4                  7         0         0         1   
1                     2                  5         0         0         0   
2                    10                  4         0         0         0   
3                     4                  6         0         0         0   
4                     6                  5         0         0         0   
...                 ...                ...       ...       ...       ...   
2235                  4                  5         0         0         0   
2236                  5                  7         1         0         0   
2237                 13                  6         1         0         0   
2238                 10                  3         0         0         0   
2239                  4                  7         0         0         1   

      ChildrenHome  
0                0  
1                2  
2                0  
3                1  
4                1  
...            ...  
2235             1  
2236             3  
2237             0  
2238             1  
2239             2  

[2240 rows x 21 columns]

In [30]:
campaigndata['Dt_Customer'] = pd.to_datetime(campaigndata['Dt_Customer'], errors='coerce')
campaigndata

Age   Education Marital_Status   Income Dt_Customer  Recency  MntWines  \
0      65  Graduation         Single  58138.0  2012-04-09       58       635   
1      68  Graduation         Single  46344.0  2014-08-03       38        11   
2      57  Graduation       Together  71613.0  2013-08-21       26       426   
3      38  Graduation       Together  26646.0  2014-10-02       26        11   
4      41         PhD        Married  58293.0  2014-01-19       94       173   
...   ...         ...            ...      ...         ...      ...       ...   
2235   55  Graduation        Married  61223.0  2013-06-13       46       709   
2236   76         PhD       Together  64014.0  2014-10-06       56       406   
2237   41  Graduation       Divorced  56981.0  2014-01-25       91       908   
2238   66      Master       Together  69245.0  2014-01-24        8       428   
2239   68         PhD        Married  52869.0  2012-10-15       40        84   

      MntFruits  MntMeatProducts  MntFishProducts  ...  MntGoldProds  \
0            88              546              172  ...            88   
1             1                6                2  ...             6   
2            49              127              111  ...            42   
3             4               20               10  ...             5   
4            43              118               46  ...            15   
...         ...              ...              ...  ...           ...   
2235         43              182               42  ...           247   
2236          0               30                0  ...             8   
2237         48              217               32  ...            24   
2238         30              214               80  ...            61   
2239          3               61                2  ...            21   

      NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0                     3                8                   10   
1                     2                1                    1   
2                     1                8                    2   
3                     2                2                    0   
4                     5                5                    3   
...                 ...              ...                  ...   
2235                  2                9                    3   
2236                  7                8                    2   
2237                  1                2                    3   
2238                  2                6                    5   
2239                  3                3                    1   

      NumStorePurchases  NumWebVisitsMonth  Previous  Complain  Response  \
0                     4                  7         0         0         1   
1                     2                  5         0         0         0   
2                    10                  4         0         0         0   
3                     4                  6         0         0         0   
4                     6                  5         0         0         0   
...                 ...                ...       ...       ...       ...   
2235                  4                  5         0         0         0   
2236                  5                  7         1         0         0   
2237                 13                  6         1         0         0   
2238                 10                  3         0         0         0   
2239                  4                  7         0         0         1   

      ChildrenHome  
0                0  
1                2  
2                0  
3                1  
4                1  
...            ...  
2235             1  
2236             3  
2237             0  
2238             1  
2239             2  

[2240 rows x 21 columns]

In [31]:
# Changing the column Dt_Customer to Year Customer
import datetime as DT

now = pd.Timestamp('now')
campaigndata.rename(columns = {'Dt_Customer':'Year_Customer'}, inplace = True)
campaigndata['Year_Customer'] = (now - campaigndata['Year_Customer']).astype('<m8[Y]') 


After scanning through the variable **Marital Status**, we see that there are multiple levels. In this case, we decide to merge `Yolo`, `Absurd`, `Widow` and `Alone` into `Single`. We do this by using the replace string method.

In [32]:
print(campaigndata["Marital_Status"].value_counts())

Married     864
Together    580
Single      480
Divorced    232
Widow        77
Alone         3
Absurd        2
YOLO          2
Name: Marital_Status, dtype: int64


In [33]:
campaigndata['Marital_Status'] = campaigndata['Marital_Status'].replace('Widow' ,'Single')
campaigndata['Marital_Status'] = campaigndata['Marital_Status'].replace('Alone' ,'Single')
campaigndata['Marital_Status'] = campaigndata['Marital_Status'].replace('YOLO' ,'Single')
campaigndata['Marital_Status'] = campaigndata['Marital_Status'].replace('Absurd' ,'Single')

In [34]:
print(campaigndata["Marital_Status"].value_counts())

Married     864
Together    580
Single      564
Divorced    232
Name: Marital_Status, dtype: int64


## Now before we do EDA, we first check whether the data contains any null entries.

In [35]:
# Checking for null values
print("Null Values:")
print(campaigndata.isnull().sum(),"\n") 

Null Values:
Age                     0
Education               0
Marital_Status          0
Income                 24
Year_Customer           0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
Previous                0
Complain                0
Response                0
ChildrenHome            0
dtype: int64 



We notice that the variable income contains null values. After much consideration, we decide to remove this 24 data entries as the number are quite small as compared to total rows. (2240 rows).

In [36]:
campaigndata = campaigndata.dropna()
campaigndata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2216 entries, 0 to 2239
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Age                  2216 non-null   int64  
 1   Education            2216 non-null   object 
 2   Marital_Status       2216 non-null   object 
 3   Income               2216 non-null   float64
 4   Year_Customer        2216 non-null   float64
 5   Recency              2216 non-null   int64  
 6   MntWines             2216 non-null   int64  
 7   MntFruits            2216 non-null   int64  
 8   MntMeatProducts      2216 non-null   int64  
 9   MntFishProducts      2216 non-null   int64  
 10  MntSweetProducts     2216 non-null   int64  
 11  MntGoldProds         2216 non-null   int64  
 12  NumDealsPurchases    2216 non-null   int64  
 13  NumWebPurchases      2216 non-null   int64  
 14  NumCatalogPurchases  2216 non-null   int64  
 15  NumStorePurchases    2216 non-null   i

In [37]:
campaigndata.to_csv("engineered_data.csv", index=False)